src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs

```diff
@@ -45,86 +45,83 @@ public HeaderModelBinder(ILoggerFactory loggerFactory)
             _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
         }
 
+        /// <summary>
+        /// Initializes a new instance of <see cref="HeaderModelBinder"/>.
+        /// </summary>
+        /// <param name="innerModelBinder">The <see cref="IModelBinder"/> which does the actual
+        /// binding of values.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public HeaderModelBinder(ILoggerFactory loggerFactory, IModelBinder innerModelBinder)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            if (innerModelBinder == null)
+            {
+                throw new ArgumentNullException(nameof(innerModelBinder));
+            }
+
+            InnerModelBinder = innerModelBinder;
+            _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
+        }
+
+        // to enable unit testing
+        internal IModelBinder InnerModelBinder { get; }
+
         /// <inheritdoc />
-        public Task BindModelAsync(ModelBindingContext bindingContext)
+        public async Task BindModelAsync(ModelBindingContext bindingContext)
         {
             if (bindingContext == null)
             {
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            var request = bindingContext.HttpContext.Request;
+            _logger.AttemptingToBindModel(bindingContext);
 
             // Property name can be null if the model metadata represents a type (rather than a property or parameter).
```

Comment:
Sounds like a corner case where the new implementation may have changed the behaviour. Do we have tests of this scenario?

Reply:
What is this actually about? Is it about decorating on a complex type? if yes, we do not support it and I have tests in this PR to check that.

Reply:
:ok: